### PR TITLE
Handle not existing forms and form files during bulk finalize

### LINKED ISSUE #5833: The app crashes if a user bulk finalizes drafts that do not have their form definitions
#### ODK Collect version
v2023.3.0

#### Problem description
The app crashes if a user bulk finalizes drafts that do not have their form definitions.

#### Steps to reproduce the problem
1. Save any form as draft
2. Go to `Settings` -> `Project management` -> `Reset`, select the third option (Blank forms...), and click the reset button.
3. Open the list of drafts and try to bulk finalize the saved forms

#### Expected behavior
There should be no crash and the draft that does not have its form definition should be ignored.

#### Other information 
It's visible in our crash reports: https://console.firebase.google.com/u/1/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/5e5e34af343a0e2172d530fdce799f08?time=last-thirty-days&types=crash&versions=v2023.3.0%20(4724)&sessionEventKey=654C458C03C9000124FF4BF5D9B31D60_1877732972590742889

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #5833

#### Why is this the best possible solution? Were any other approaches considered?
Nothing to discuss here.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This should only fix the crashes mentioned in the issue and have no side effects. I can't think of any risk here.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] added a comment above any new strings describing it for translators
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by lognaturel** [comment]
It looks like this will show the finalization attempt as failed but won't give any context as to why, is that right? The user won't have any hints about what happened or how they could fix it? Is that better than just crashing?

**Comment by grzesiek2010** [comment]
> It looks like this will show the finalization attempt as failed but won't give any context as to why, is that right?

Correct.

> the user won't have any hints about what happened or how they could fix it? Is that better than just crashing?

To me of course it's better:
1. The message displayed after reopening the app wouldn't be helpful at all:
image

2. If you have many filled forms and only one or some of them can't be finalized because of this issue then still the rest of the forms will be finalized. If you have a crash the process is stopped.

3. A user can try to open non-finalized forms and then see this, which is helpful:
image

File: collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryUseCases.kt [hunk 1]
```diff
@@ -29,21 +29,28 @@ object FormEntryUseCases {
         formsRepository: FormsRepository,
         projectRootDir: File,
         formDefCache: FormDefCache
-    ): Pair<FormDef, Form> {
+    ): Pair<FormDef?, Form?> {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryUseCases.kt]: Could this return `Pair<FormDef, Form>?` instead and then the check on the other side could just be if the whole result is null?

**Review by seadowg — CHANGES_REQUESTED** [review]

**Comment by lognaturel** [comment]
Great, I didn’t think through what they’d see if they try to open the form. Makes sense!

File: collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryUseCases.kt (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryUseCases.kt]: I was thinking about that but thought it wouldn't be better... but maybe https://github.com/getodk/collect/pull/5835/commits/fed5b7de5134201df0bcbf6651885d993fa79b31
If you like it more we can keep it.

**Review by seadowg — APPROVED** [review]

**Comment by dbemke** [comment]
On Android 10 and 13 after reproducing steps from the issue there is a crash.

    Save any form as draft.
    Delete the form file manually from the project directory (the one from the forms subdirectory)
    Open the list of drafts and try to bulk finalize the saved forms

**Comment by grzesiek2010** [comment]
> On Android 10 and 13 after reproducing steps from the issue there is a crash.

What exactly are you removing? Only the XML file or something else? I've just tried and everything seems to be fine.
Does it only occur on Android 10 and 13 or you have just tried those two devices?

**Comment by srujner** [comment]
I removed just the Xml document from the Files -> projects -> Name of the project -> instances -> Name of the form.

I tried only on Android 13 so far

**Comment by grzesiek2010** [comment]
> I removed just the Xml document from the Files -> projects -> Name of the project -> instances -> Name of the form.

That's a completely new case different than those two I've described in the issue. But ok we can handle this one too.

**Comment by srujner** [comment]
So what's the one mention in here: 
"Delete the form file manually from the project directory (the one from the forms subdirectory)" 

Deleting the whole folder?

**Comment by grzesiek2010** [comment]
> So what's the one mention in here:
"Delete the form file manually from the project directory (the one from the forms subdirectory)"

> Deleting the whole folder?

No, the difference is that I said: delete the file from the **forms** subdirectory and you removed the one from the **instances** subdirectory. So instead of removing the blank form file, you removed the filled form file.

**Comment by dbemke** [comment]
We reproduced the steps from the issue with resetting blank forms and the saved drafts has "No errors" pill but after trying to bulk finalize there's the snackbar with information that the form contains errors. Is it expected?
Noerrorsdeletedblanks

**Comment by dbemke** [comment]
Is it the expected error message after removing the form subdirectory and trying to open the saved instance of the form?

deletedFormSubdirectory

**Comment by grzesiek2010** [comment]
> We reproduced the steps from the issue with resetting blank forms and the saved drafts has "No errors" pill but after trying to bulk finalize there's the snackbar with information that the form contains errors. Is it expected?

Yes, those are edge cases and not errors in forms themselves so it's ok as it is.

> Is it the expected error message after removing the form subdirectory and trying to open the saved instance of the form?

I would say yes. The message could be better maybe but is not that bad if a user reads it, it's possible to figure out the reason.

**Comment by srujner** [comment]
**Tested with Success!**

Verified on device with Android 13

Verified cases:
- Issue #5833 is no longer reproducing;
- Removing files from Forms/Instance folders;
- Regression check on Bulk Finalize;
- Editing Drafts;
- Regression check on Project Management Settings;
- Don't keep activities enabled/disabled;
- Rotate the screen, minimize the app;
- Light and Dark mode

**Comment by dbemke** [comment]
**Tested with Success!**

Verified on device with Android 10

**Pull request merged**